compiler/rustc_codegen_ssa/src/base.rs

```diff
@@ -273,12 +273,11 @@ pub fn cast_to_dyn_star<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
         matches!(dst_ty.kind(), ty::Dynamic(_, _, ty::DynStar)),
         "destination type must be a dyn*"
     );
-    // FIXME(dyn-star): this is probably not the best way to check if this is
-    // a pointer, and really we should ensure that the value is a suitable
-    // pointer earlier in the compilation process.
-    let src = match src_ty_and_layout.pointee_info_at(bx.cx(), Size::ZERO) {
-        Some(_) => bx.ptrtoint(src, bx.cx().type_isize()),
-        None => bx.bitcast(src, bx.type_isize()),
+    let src = match bx.cx().type_kind(bx.cx().backend_type(src_ty_and_layout)) {
+        TypeKind::Pointer => bx.pointercast(src, bx.cx().type_i8p()),
+        TypeKind::Integer => bx.inttoptr(src, bx.cx().type_i8p()),
+        // FIXME(dyn-star): We probably have to do a bitcast first, then inttoptr.
+        kind => bug!("unexpected TypeKind for left-hand side of `dyn*` cast: {kind:?}"),
```

Comment:
This can probably be hit by using a type such as
```rust
#[repr(C)]
struct S(usize);
```
and then doing `S(0) as dyn* Send` or something like that.

Reply:
That's broken for another reason (#104694), which I plan on rebasing/reworking once your Miri support for dyn* PR lands.

Reply:
Yeah that's not surprising -- I assume so far this has only been tested with types that have Scalar layout.

`(i32, i32)` would also be an interesting testcase, that will have ScalarPair layout. Also better left for a future PR, probably.